Modules/_abc.c

```diff
@@ -527,10 +531,14 @@ _abc__abc_register_impl(PyObject *module, PyObject *self, PyObject *subclass)
     /* Invalidate negative cache */
     get_abc_state(module)->abc_invalidation_counter++;
 
+    /* Set Py_TPFLAGS_SEQUENCE  or Py_TPFLAGS_MAPPING flag */
     if (PyType_Check(subclass) && PyType_Check(self) &&
         !PyType_HasFeature((PyTypeObject *)subclass, Py_TPFLAGS_IMMUTABLETYPE))
     {
-        ((PyTypeObject *)subclass)->tp_flags |= (((PyTypeObject *)self)->tp_flags & COLLECTION_FLAGS);
+        if (((PyTypeObject *)self)->tp_flags & COLLECTION_FLAGS) {
+            ((PyTypeObject *)subclass)->tp_flags &= ~COLLECTION_FLAGS;
+            ((PyTypeObject *)subclass)->tp_flags |= (((PyTypeObject *)self)->tp_flags & COLLECTION_FLAGS);
+        }
     }
```

Comment:
This might read better if the nested "if" conditions are merged. Also (looking at this again), I don't think the first `PyType_Check` call is needed.

So maybe something like:

```c
    /* Set Py_TPFLAGS_SEQUENCE or Py_TPFLAGS_MAPPING flag */
    if (PyType_Check(self) &&
        PyType_HasFeature((PyTypeObject *)self, COLLECTION_FLAGS) &&
        !PyType_HasFeature((PyTypeObject *)subclass, Py_TPFLAGS_IMMUTABLETYPE))
    {
        ((PyTypeObject *)subclass)->tp_flags &= ~COLLECTION_FLAGS;
        ((PyTypeObject *)subclass)->tp_flags |= (((PyTypeObject *)self)->tp_flags & COLLECTION_FLAGS);
    }
```

(GitHub won't let me suggest changing lines that aren't part of the diff.)